### DISCUSSION TITLE
Get OsgEarth Texture and render to a GLFW/IMGUI window

### DISCUSSION THREAD

**Discussion by Mattc2021** [pelicanmapping/osgearth#2230] [Support Q&A] [answered]
Was just curious what the best way to render the texture of osgEarth to a camera and then be able to see it in an imgui window. Currently my code for setting up the RTT camera looks like this...

osg::ref_ptr<osg::Image> image = new osg::Image;
m_texture = new osg::Texture2D;
osg::ref_ptr<osg::Camera> camera = createRTTCamera(1600, 900, m_texture.get(), image);
//camera->addChild(scene);
m_viewer->setCamera(camera);
m_viewer->setUpViewerAsEmbeddedInWindow(0, 0, 1600, 900);

with the createRTTCamera looking like this

osg::Camera* OSGEarthViewer::createRTTCamera(int width, int height, osg::ref_ptr<osg::Texture2D> texture, osg::ref_ptr<osg::Image> image)
    {
        texture->setTextureSize(width, height);
        texture->setInternalFormat(GL_RGB);
        texture->setFilter(osg::Texture2D::MIN_FILTER, osg::Texture2D::LINEAR);
        texture->setFilter(osg::Texture2D::MAG_FILTER, osg::Texture2D::LINEAR);
        osg::ref_ptr<osg::Camera> rttCamera = new osg::Camera;
        rttCamera->setClearMask(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
        rttCamera->setClearColor(osg::Vec4(1.0f, 1.0f, 1.0f, 1.0f));
        rttCamera->setViewport(0, 0, texture->getTextureWidth(), texture->getTextureHeight());
        rttCamera->setRenderOrder(osg::Camera::PRE_RENDER);
        rttCamera->setReferenceFrame(osg::Transform::RELATIVE_RF);
        rttCamera->setRenderTargetImplementation(osg::Camera::FRAME_BUFFER_OBJECT);
        rttCamera->attach(osg::Camera::COLOR_BUFFER, texture);
        rttCamera->attach(osg::Camera::COLOR_BUFFER, image);
        return rttCamera.release();
    }

Then in my render loop I'm doing 
osg::Texture::TextureObject* tex_obj = osgViewerTexture->getTextureObject(context_id);
    if (tex_obj) {
        ImGui::Begin("Window A");
        ImGui::Image((void*)(tex_obj->id()), ImVec2{ 1600, 900 });
        ImGui::End(); 
    }

finally I'm simply calling frame on my viewer to return it. Any help would be appreciated as it's just coming back with a white screen. Also side note, the tex_obj has a texture associated with it, but doesn't have an image attached to it.

**Comment by gwaldron** [comment, marked as answer]
This utility function might help:
https://github.com/gwaldron/osgearth/blob/master/src/osgEarth/ImGui/ImGui#L249
In particular, the texture->apply() is probably necessary.

**Reply by Mattc2021** [reply]
Thank you so much! Hoping this helps, but when I see the RenderInfo is that set by the viewer and I access it through there or is it through the GraphicsContext that I get from my camera.

**Reply by gwaldron** [reply]
Either way should work. You should be able to call GraphicsContext::getState() as well.
